File: 2021/Day03.java
public class Day03 {

  public static void main(String[] args) {
String[] binary = {
};
boolean[][] valueBase2 = new boolean[binary.length][binary[1].length()];
    for (int i = 0; i < binary.length; i++) {
      for(int j = 0; j <binary[i].length(); j++){
        if (binary[i].charAt(j) == '1'){
          valueBase2[i][j] = true;
        }
      }
    }
    // for (int i = 0; i < valueBase2.length; i++) {
    //   for (int j = 0; j < valueBase2[i].length; j++ ) {
    //     if (valueBase2[i][j] == true) {
    //       System.out.print(1);
    //     }
    //     else{
    //       System.out.print(0);
    //     }
    //   }
    //   System.out.println();
    // }



    boolean[] gamma = new boolean[binary[1].length()];
    boolean[] epsilon = new boolean[gamma.length];
    for (int j = 0; j < gamma.length ; j++) {
      int counttrue = 0;
      int countfalse = 0;
      for(int i = 0; i < valueBase2.length; i++){
        if(valueBase2[i][j] == true){
          counttrue = counttrue + 1;
        }
        else{
          countfalse = countfalse + 1;
        }
      }
      if(counttrue >= countfalse){
          gamma[j] = true;

      }
      if(countfalse >= counttrue){
        epsilon[j] = true;

      }

    }

    for (int i = 0; i < gamma.length; i++ ) {
      if (gamma[i] == true) {
        System.out.print(1);
      }
      else{
        System.out.print(0);
      }
    }
    System.out.println();
    for (int i = 0; i < epsilon.length; i++ ) {
      if (epsilon[i] == true) {
        System.out.print(1);
      }
      else{
        System.out.print(0);
      }
    }
    System.out.println();
    int gammadez = 0;
    for(int i = 0; i< gamma.length; i++){
      gammadez = 2 * gammadez;
      if(gamma[i] == true){
        gammadez = gammadez + 1;
      }
    }
    System.out.println(gammadez);
    int epsilondez = 0;
    for(int i = 0; i< epsilon.length; i++){
      epsilondez = 2 * epsilondez;
      if(epsilon[i] == true){
        epsilondez = epsilondez + 1;
      }
    }
    System.out.println(epsilondez);
    System.out.println(epsilondez * gammadez);
  }
}

File: 2021/Day03.py
input = open('Day03-imput.txt', 'r')

def p1(input):
    lines = input.readlines()
    for k in range(len(lines)):
        lines[k] = list(lines[k].replace('\n', ''))
    gamma = []
    for l in range(len(lines[0])):
        zero, one = 0, 0
        for k in range(len(lines)):
            if lines[k][l] == '0':
                zero += 1
            else:
                one += 1
        if zero > one:
            gamma.append('0')
        else:
            gamma.append('1')
    epsilon = []
    for k in range(len(gamma)):
        if gamma[k] == '0':
            epsilon.append('1')
        else:
            epsilon.append('0')
    str1 = ''
    str2 = ''
    for k in range(len(gamma)):
        str1 += gamma[k]
        str2 += epsilon[k]
    print(int(str1, 2) * int(str2, 2))


def p2(input):
    lines = input.readlines()
    # print(lines)
    for k in range(len(lines)):
        lines[k] = list(lines[k].replace('\n', ''))
    lines.sort()
    co2lines = lines.copy()
    o2lines = lines.copy()
    co2value = []
    o2value = []
    for l in range(12):
        one, zero = 0, 0
        for k in range(len(o2lines)):
            if o2lines[k][l] == '0':
                zero += 1
            else:
                one += 1
        if one >= zero:
            val = '1'
            o2value.append('1')
        else:
            val = '0'
            o2value.append('0')
        newo2 = []
        for k in range(len(o2lines)):
            if o2lines[k][l] == val:
                newo2.append(o2lines[k])
            else:
                pass
        o2lines = newo2
        if len(o2lines) == 1:
            o2value = o2lines[0]
            break
    for l in range(12):
        one, zero = 0, 0
        for k in range(len(co2lines)):
            if co2lines[k][l] == '0':
                zero += 1
            else:
                one += 1
        if zero <= one:
            val = '0'
            o2value.append('0')
        else:
            o2value.append('1')
            val = '1'
        newco2 = []
        for k in range(len(co2lines)):
            if co2lines[k][l] == val:
                newco2.append(co2lines[k])
            else:
                pass
        co2lines = newco2
        if len(co2lines) == 1:
            co2value = co2lines[0]
            break
    ## Why o2value gets longer then 12 bits is still a mystery to me, but this gives the right result...
    print(int(''.join(co2value), 2) * int(''.join(o2value[0:12]), 2))




p2(input)
